File: src/common_basis.py
import keyring
import getpass
import sqlalchemy
import os
import yaml
from hereutil import here
from sqlalchemy import text
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.future import Engine, Connection


def get_connection() -> (Engine, Connection):
    eng = None
    con = None
    while con is None:
        with here("db_params.yaml").open('r') as f:
            db_params = yaml.safe_load(f)
        password = os.environ.get("DB_PASS")
        if password is None:
            try:
                password = keyring.get_password(db_params['db_name'], "DB_PASS")
            except keyring.errors.NoKeyringError:
                pass
        try:
            eng = sqlalchemy.create_engine(
                "mariadb+pymysql://" + db_params['db_user'] + ":" + password + "@" + db_params['db_host'] + "/" +
                db_params['db_name'] + "?charset=utf8mb4&autocommit&local_infile",
                future=True
            )
            con = eng.connect()
        except SQLAlchemyError as err:
            eng = None
            con = None
            password = getpass.getpass(f"Database password (connection attempt failed with {err}): ")
            try:
                if keyring.get_password(db_params['db_name'], "DB_PASS") is not None:
                    keyring.delete_password(db_params['db_name'], "DB_PASS")
                keyring.set_password(db_params['db_name'], "DB_PASS", password)
            except keyring.errors.NoKeyringError:
                pass
    return eng, con


eng, con = get_connection()
del get_connection


def set_session_storage_engine(con: Connection, engine: str):
    con.execute(text("SET SESSION storage_engine="+engine))


__all__ = ["eng", "con", "set_session_storage_engine"]

File: src/common_basis.R
library(here)
library(tidyverse)
library(hscidbutil)
library(keyring)
library(DBI)
library(RMariaDB)
library(googlesheets4)
library(yaml)

get_connection <- function() {
  while (!exists("con")) {
    db_params <- read_yaml(here("db_params.yaml"))
    tryCatch(con <- dbConnect(
      drv = MariaDB(),
      host = db_params$db_host,
      dbname = db_params$db_name,
      user = db_params$db_user,
      password = if (file.exists(here("db_secret.yaml"))) read_yaml(here("db_secret.yaml"))$db_pass else key_get(db_params$db_name,"DB_PASS"),
      bigint = "integer",
      load_data_local_infile = TRUE,
      autocommit = TRUE,
      reconnect = TRUE
    ), error = function(e) {
      print(e)
      key_set(db_params$db_name,"DB_PASS", prompt="Database password: ")
    })
  }
  con
}

con <- get_connection()
rm(get_connection)

try(test_a <- tbl(con, "test_a"))
try(test_c <- tbl(con, "test_c"))

